designs/2019-variable-definition-information-of-config-files/README.md

```diff
@@ -0,0 +1,86 @@
+- Start Date: 2019-03-04
+- RFC PR: (leave this empty, to be filled in later)
+- Authors: Toru Nagashima ([@mysticatea])
+
+# Variable Definition Information of Config Files
+
+## Summary
+
+This proposal adds variable definition information of config files to `Variable` objects.
+
+## Motivation
+
+To make rules being able to report matters about variable definitions related to config files.
+
+Especially, this proposal minds [no-redeclare] rule to report redeclaration by a mix of regular variable declarations, `/*globals*/` comments, and config files. ([eslint/eslint#11370])
+
+## Detailed Design
+
+Currently, `Variable` objects have the following properties to express `/*globals*/` definitions.
+
+- `variable.eslintExplicitGlobal` (`boolean`) ... `true` if this variable was defined by `/*globals*/` comment.
+- `variable.eslintExplicitGlobalComment` (`Comment` object) ... The `/*globals*/` comment that defines this variable. If two or more comments defined this variable, it adopts the last one.
+- `variable.writeable` (`boolean`) ... `true` if the final setting is `"writable"`.
+
+This proposal adds the following properties.
+
+- `variable.eslintExplicitGlobalComments` (an array of `Comment` objects) ... All `/*globals*/` comments that define this variable.
+- `variable.eslintImplicitGlobalSetting` (`"readonly"` | `"writable"` | `null`) ... The setting in config files. This is `null` if config files didn't define it.
+
+Both `variable.eslintExplicitGlobalComments` and `variable.eslintImplicitGlobalSetting` are instantiated even if regular variable declarations (E.g. `let foo` syntax) exist. Therefore, [no-redeclare] rule can verify redeclaration by a mix of regular variable declarations, `/*globals*/` comments, and config files.
+
+This will be implemented into [`addDeclaredGlobals(globalScope, configGlobals, commentDirectives)`][lib/linter.js#L73-L131] function.
+
+<table><td>
+<a id="question1">❓</a> <b>Open Question</b>:<br>
+(<a href="question1a">link</a>) Should we handle de-declaration such as <code>/*globals foo:off*/</code>? Because variable objects don't exist for de-declared variables, rules cannot know de-declared variables under this proposal.
+</td></table>
+
+## Documentation
+
+All of the existing `variable.eslintExplicitGlobal`, `variable.eslintExplicitGlobalComment`, and `variable.writeable` are undocumented. So new properties also can be undocumented.
+
+<table><td>
+<a id="question2">❓</a> <b>Open Question</b>:<br>
+(<a href="question2a">link</a>) Should we document those properties as public API officially?
```

Comment:
I think we should document the new properties, at least. (It seems like `variable.writeable` is an important feature, so I'm surprised that it's undocumented.)